### PR TITLE
[clang] Use separator for large numeric values in overflow diagnostic

### PR DESCRIPTION
**Pull request by Atousa**
Add functionality to APInt::toString() that allows it to insert
separators between groups of digits, using the C++ litteral
separator ' between groups.
    
Fixes issue #58228

Reviewers :  @AaronBallman, @cjdb, @tbaederr

### CODE REVIEW

**Comment by AaronBallman** [comment]
Overall, I think this is a *fantastic* idea for diagnostic readability, thank you for the patch! Please be sure to add a release note so users known to expect the changes. The Windows precommit CI failure is unrelated and can be ignored.

Once you think you're ready, be sure to move the PR out of draft status so reviewers remember to give it a proper review. My initial glance didn't spot any issues though.

File: llvm/lib/Support/APInt.cpp [hunk 1]
```diff
@@ -2187,6 +2188,12 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
     }
   }
 
+  // Number of digits in a group between separators.
+  unsigned Grouping = (Radix == 8 || Radix == 10) ? 3 : 4;
+  if (Radix == 8 || Radix == 10) {
+    Grouping = 3;
+  }
+
```
Comment (tbaederr) [inline llvm/lib/Support/APInt.cpp:2196]: This code block is now unnecessary.

**Comment by Atousa** [comment]
Humble ping @cjdb 
Do you have any further comments? @AaronBallman @tbaederr

**Comment by Atousa** [comment]
@AaronBallman @tbaederr, Do you have any further comments?

File: llvm/include/llvm/ADT/APInt.h [hunk 2]
```diff
@@ -1626,7 +1626,8 @@ class [[nodiscard]] APInt {
   /// SmallString. If Radix > 10, UpperCase determine the case of letter
   /// digits.
   void toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
-                bool formatAsCLiteral = false, bool UpperCase = true) const;
+                bool formatAsCLiteral = false, bool UpperCase = true,
+                bool insertSeparators = false) const;
```
Comment (AaronBallman) [inline llvm/include/llvm/ADT/APInt.h:1630]:
```suggestion
                bool InsertSeparators = false) const;
```
This matches our usual coding style and is the prevailing style in the function signature (a follow-up NFC change could rename `formatAsCLiteral`).

File: llvm/include/llvm/ADT/StringExtras.h [hunk 3]
```diff
@@ -329,9 +329,10 @@ inline std::string itostr(int64_t X) {
 }
 
 inline std::string toString(const APInt &I, unsigned Radix, bool Signed,
-                            bool formatAsCLiteral = false) {
+                            bool formatAsCLiteral = false,
+                            bool upperCase = true, bool addSeparators = false) {
```
Comment (AaronBallman) [inline llvm/include/llvm/ADT/StringExtras.h:333]: Same suggestion here as above (amusing that `upperCase` changed styles...).

File: llvm/lib/Support/APInt.cpp [hunk 4]
```diff
@@ -2161,7 +2161,8 @@ void APInt::fromString(unsigned numbits, StringRef str, uint8_t radix) {
 }
 
 void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
-                     bool formatAsCLiteral, bool UpperCase) const {
+                     bool formatAsCLiteral, bool UpperCase,
+                     bool insertSeparators) const {
```
Comment (AaronBallman) [inline llvm/lib/Support/APInt.cpp:2165]: And here too!

File: llvm/lib/Support/APInt.cpp [hunk 5]
```diff
@@ -2257,17 +2266,27 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
     unsigned ShiftAmt = (Radix == 16 ? 4 : (Radix == 8 ? 3 : 1));
     unsigned MaskAmt = Radix - 1;
 
+    int Pos = 0;
     while (Tmp.getBoolValue()) {
       unsigned Digit = unsigned(Tmp.getRawData()[0]) & MaskAmt;
+      if (insertSeparators && Pos % Grouping == 0 && Pos > 0) {
+        Str.push_back('\'');
+      }
```
Comment (AaronBallman) [inline llvm/lib/Support/APInt.cpp:2274]: Same here as above

File: llvm/lib/Support/APInt.cpp [hunk 6]
```diff
@@ -2257,17 +2266,27 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
...
     }
   } else {
+    int Pos = 0;
     while (Tmp.getBoolValue()) {
       uint64_t Digit;
       udivrem(Tmp, Radix, Tmp, Digit);
       assert(Digit < Radix && "divide failed");
+      if (insertSeparators && Pos % Grouping == 0 && Pos > 0) {
+        Str.push_back('\'');
+      }
```
Comment (AaronBallman) [inline llvm/lib/Support/APInt.cpp:2287]: Here too

File: llvm/unittests/ADT/APIntTest.cpp [hunk 7]
```diff
@@ -1379,6 +1379,23 @@ TEST(APIntTest, toString) {
...
+  EXPECT_EQ(std::string(S), "02'000");
+  S.clear();
+  APInt(64, 1000000).toString(S, 10, false, true, false, true);
+  EXPECT_EQ(std::string(S), "1'000'000");
+  S.clear();
+  APInt(64, 1000000).toString(S, 16, false, true, true, true);
+  EXPECT_EQ(std::string(S), "0xF'4240");
+  S.clear();
+  APInt(64, 1'000'000'000).toString(S, 36, false, false, false, true);
+  EXPECT_EQ(std::string(S), "gj'dgxs");
```
Comment (AaronBallman) [inline llvm/unittests/ADT/APIntTest.cpp:1396]: lol, I love it (it took me a minute when I saw the expected string).

**Review by AaronBallman — COMMENTED** [review]
You should also add a release note to `clang/docs/ReleaseNotes.rst` to tell users about the improved diagnostic behavior.

Aside from some small nits, this looks really close to ready!

**Comment by Atousa** [comment]
please @AaronBallman

File: clang/docs/ReleaseNotes.rst [hunk 8]
```diff
@@ -175,6 +175,9 @@ Improvements to Clang's diagnostics
 - The ``-Wshorten-64-to-32`` diagnostic is now grouped under ``-Wimplicit-int-conversion`` instead
    of ``-Wconversion``. Fixes `#69444 <https://github.com/llvm/llvm-project/issues/69444>`_.
 
+- Clang now uses thousand separators when printing large numbers in integer overflow diagnostics.
+  Fixes `#80939 <https://github.com/llvm/llvm-project/issues/80939>`_.
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:179]:
```suggestion
  Fixes #GH80939.
```
We just landed a change this morning to automagically expand GitHub issue numbers, so this isn't a functional change.

File: clang/lib/AST/ExprConstant.cpp [hunk 9]
```diff
@@ -2778,7 +2778,9 @@ static bool CheckedIntArithmetic(EvalInfo &Info, const Expr *E,
     if (Info.checkingForUndefinedBehavior())
       Info.Ctx.getDiagnostics().Report(E->getExprLoc(),
                                        diag::warn_integer_constant_overflow)
-          << toString(Result, 10) << E->getType() << E->getSourceRange();
+          << toString(Result, 10, Result.isSigned(), /*formatAsCLiteral=*/false,
+                      /*upperCase=*/true, /*insertSeparators=*/true)
```
Comment (AaronBallman) [inline clang/lib/AST/ExprConstant.cpp:2782]:
```suggestion
                      /*UpperCase=*/true, /*InsertSeparators=*/true)
```
Same changes needed elsewhere, but this should keep clang-tidy happy for folks using it.

**Review by AaronBallman — APPROVED** [review]
LGTM, thank you for the improvement!

**Pull request merged**